i2c_software: allow freq adjust

nefelim4ag opened the pull request:
Just a follow-up, from SPI SW #6856 and ICM20948 fixes #6868.
Time logic is imported from SW SPI.

Adjustable I2C, the maximal frequency that I was able to achieve with STM32H7 is ~800kHz.
Req | Achived
-- | -- 
1000kHz | 460kHz
800kHz | 416kHz
600kHz | 354kHz
400kHz | 273kHz
200kHz | 162Khz
100kHz | 90kHz

_I suspect the real limit here, is the gpio_in/out_reset with `gpio_peripheral()`._

Thanks.

nefelim4ag committed:
i2c_software: allow freq adjust

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

MRX8024 commented:
I tested this patch with the same STM32H7 and got max ~[700khz ](https://github.com/user-attachments/assets/06e9b54f-9c20-4ae8-a591-74d07702c308)rate.

KevinOConnor commented:
Thanks for working on this.  We can certainly improve the i2c software code.  Same comments as #6856 though - the host should pass in the delay ticks, the delay needs to verify each pulse length, and this should go in after v0.13.0 is tagged.

Also note #6885 - I'm inclined to commit that before v0.13.0 as it looks like there is a counter rollover bug in the current code.

-Kevin

nefelim4ag committed:
i2c_software: pass pulse ticks from host

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag committed:
i2c_software: reduce gpio calls count

gpio reset calls are heavy.
gpio state are persistent between calls.
Drop useless calls.
Avoid calls if SDA does not changed.

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

KevinOConnor commented:
Thanks.  Looks good to me.  I think it something we should look at merging after tagging v0.13.0 .

-Kevin

KevinOConnor commented:
Thanks.

-Kevin
